### ISSUE TITLE
Nested function calls have ) token in the wrong place

### ISSUE THREAD

**Issue by vjeux** [shellscape/postcss-values-parser#26]
When parsing

```js
a(b(c(d())))
```

The `)` isn't placed in the proper place: the last three `)` tokens are placed in the before-last function node instead of the corresponding opening.

```js
Root {
  type: 'root',
  nodes:
   [ Value {
       type: 'value',
       nodes:
        [ FunctionNode {
            type: 'func',
            value: 'a',
            nodes:
             [ Parenthesis { type: 'paren', value: '(' },
               FunctionNode {
                 type: 'func',
                 value: 'b',
                 nodes:
                  [ Parenthesis { type: 'paren', value: '(' },
                    FunctionNode {
                      type: 'func',
                      value: 'c',
                      nodes:
                       [ Parenthesis { type: 'paren', value: '(' },
                         FunctionNode {
                           type: 'func',
                           value: 'd',
                           nodes:
                            [ Parenthesis { type: 'paren', value: '(' },
                              Parenthesis { type: 'paren', value: ')' } ] },
                         Parenthesis { type: 'paren', value: ')' },
                         Parenthesis { type: 'paren', value: ')' },
                         Parenthesis { type: 'paren', value: ')' } ] } ] } ] } ] } ] }
```

Program to reproduce:

```js
const parser = require('postcss-values-parser')
const str = 'a(b(c(d())))';
const value = parser(str).parse();

log(clean(value));

function log(node) {
  console.log(require('util').inspect(node, false, null));
}

function clean(node) {
  if (Array.isArray(node)) {
    node.forEach(clean);
  } else if (node && typeof node === "object") {
    const type = node.type;
    const value = node.value;
    const nodes = node.nodes;
    for (let key in node) {
      if (key !== "parent") {
        clean(node[key]);
      }
      delete node[key];
    }
    if (type) node.type = type;
    if (value) node.value = value;
    if (nodes) node.nodes = nodes;
  }
  return node;
}
```

For context, I'm trying to use this to build a CSS pretty printer: https://github.com/prettier/prettier/pull/1636

**Comment by vjeux** [comment]
I think that the issue is on this line: https://github.com/shellscape/postcss-values-parser/blob/master/lib/parser.js#L265-L271 it requires a stack instead of a single value.

**Comment by vjeux** [comment]
With this patch it gets it correct but for the outermost one:

```diff
--- a/lib/parser.js
+++ b/lib/parser.js
@@ -29,6 +29,7 @@ module.exports = class Parser {
   constructor (input, options) {
     const defaults = { loose: false };

+    this.stack = [];
     this.input = input;
     this.options = Object.assign({}, defaults, options);
     this.position = 0;
@@ -266,7 +267,7 @@ module.exports = class Parser {

     if (last && last.type === 'func' && last.unbalanced < 0) {
       last.unbalanced = 0; // ok we're ready to add parens now
-      this.cache = this.current;
+      this.stack.push(this.current);
       this.current = last;
     }

@@ -355,9 +356,8 @@ module.exports = class Parser {
       this.error('Expected opening parenthesis', token);
     }

-    if (!this.current.unbalanced && this.cache) {
-      this.current = this.cache;
-      this.cache = null;
+    if (!this.current.unbalanced && this.stack.length) {
+      this.current = this.stack.pop();
     }
   }
```

```js
Root {
  type: 'root',
  nodes:
   [ Value {
       type: 'value',
       nodes:
        [ FunctionNode {
            type: 'func',
            value: 'a',
            nodes:
             [ Parenthesis { type: 'paren', value: '(' },
               FunctionNode {
                 type: 'func',
                 value: 'b',
                 nodes:
                  [ Parenthesis { type: 'paren', value: '(' },
                    FunctionNode {
                      type: 'func',
                      value: 'c',
                      nodes:
                       [ Parenthesis { type: 'paren', value: '(' },
                         FunctionNode {
                           type: 'func',
                           value: 'd',
                           nodes:
                            [ Parenthesis { type: 'paren', value: '(' },
                              Parenthesis { type: 'paren', value: ')' } ] },
                         Parenthesis { type: 'paren', value: ')' } ] },
                    Parenthesis { type: 'paren', value: ')' },
                    Parenthesis { type: 'paren', value: ')' } ] } ] } ] } ] }
```

**Comment by shellscape** [comment]
Thank you for your work looking into this, but please note that Issue Templates are _not_ optional. Please fill in the missing info.

We have a pending fix here: https://github.com/shellscape/postcss-values-parser/tree/fix-nested-funcs please test that branch against your test case.

**Comment by shellscape** [comment]
Fixed in f4a7896fe44fd611c64803e0b06cd5709230f3b4
